### PR TITLE
Chore: Use KVStore for the manifest public key

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

The goal of this PR is to evaluate if we want to use the KVStore to keep track of the manifest public key. The changes in this PR are:

 - Drop the values in memory in favor of using KVStorage
 - If the storage is empty, pre-populate it with the hardcoded key (migration behavior). This means that there is no call to grafana.com by default.
 - If a plugin keyID is not present in the local storage, download API keys.
 - Starts a go routine to download the API keys every 10 days.

Current problems:

 - Only if the server process is alive for 10 days the API keys will be rotated.
 - There is no way to check how old an entry in the KV storage is (at least with the given API), so we could be relying on old entries.

I believe this could be mitigated if an initial go routine downloads the API keys without blocking the main process (e.g. the first load would use the hardcoded/old keys but the next one would use a fresh key).

**Which issue(s) does this PR fix?**:

Ref #23619

**Special notes for your reviewer:**

Thoughs? +1 -1? Alternatives? cc/ @DanCech

### CODE REVIEW

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 1]
```diff
@@ -30,17 +34,30 @@ type ManifestVerifier struct {
...
+}
+
+func New(cfg *config.Cfg, mlog log.Logger, kv *kvstore.NamespacedKVStore) *ManifestVerifier {
+	pmv := &ManifestVerifier{
+		cfg:  cfg,
+		mlog: mlog,
+		cli:  makeHttpClient(),
+		kv:   kv,
+	}
+	pmv.updatePeriodically()
```
Comment (andresmgot) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:49]: TODO: put this line behind the feature flag

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 2]
```diff
@@ -111,20 +112,68 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+		return errors.New("missing public key")
 	}
 
 	for _, key := range data.Items {
-		pmv.publicKeys[key.KeyID] = key
+		err = pmv.kv.Set(ctx, key.KeyID, key.PublicKey)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (DanCech) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:127]: we may want to use GetAll to grab all current keys and clean up any keys that exist in the store but aren't in Items, so that any bad entries are removed from the database automatically

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 3]
```diff
@@ -111,20 +112,68 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+
+	key, exist, err := pmv.kv.Get(ctx, keyID)
+	if err != nil {
+		return "", err
+	}
+	if exist {
+		return key, nil
+	}
+
+	// Not in the cache, download keys
```
Comment (DanCech) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:165]: it may be worth keeping track of the last time the keys were downloaded so that a plugin with a bad signature doesn't constantly trigger re-downloading the keys.

File: pkg/plugins/manager/signature/manifestverifier/verifier.go (hunk 2 above)
Comment (DanCech) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go:127]: If we decide to keep track of the last download time, the keys should likely be prefixed (eg with key-) so we can use Keys to get all keys and still be able to store a last updated date.

When we have last updated date we can and should also use that to bootstrap the background update process so that instances that don't run for 10 days continuously will still update their keys after 10 days.  There is an example of how to do that using the kvstore here https://github.com/grafana/grafana/blob/main/pkg/infra/usagestats/service/service.go#L61-L99

Comment (andresmgot) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go:196]: done both things :+1:

File: pkg/plugins/manager/signature/manifestverifier/verifier.go (hunk 3 above)
Comment (andresmgot) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go]: Removed this synchronous download, now it always happen in parallel.

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 4]
```diff
@@ -111,20 +174,73 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
-		return "", err
+		return err
 	}
 
 	if len(data.Items) == 0 {
-		return "", errors.New("missing public key")
+		return errors.New("missing public key")
 	}
 
+	cachedKeys, err := pmv.kv.Keys(ctx, "")
```
Comment (DanCech) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:184]: this is going to include the last_updated item, meaning that it (and any other extra k/v pairs in the namespace) will get deleted in the block below, we should prefix the k/v key for the keys so we can target them accurately.
```suggestion
	cachedKeys, err := pmv.kv.Keys(ctx, "key-")
```

Comment (andresmgot) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go:184]: right, I missed this, thanks!

File: pkg/plugins/manager/signature/manifest.go [hunk 5]
```diff
@@ -17,6 +17,7 @@ import (
 
 	"github.com/ProtonMail/go-crypto/openpgp/clearsign"
 	"github.com/gobwas/glob"
+	"github.com/grafana/grafana/pkg/infra/kvstore"
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/plugins/manager/signature/manifest.go]: done, created a new interface and service for this :+1:

File: pkg/plugins/manager/loader/loader_test.go [hunk 6]
```diff
@@ -1110,115 +1107,6 @@ func TestLoader_Load_SkipUninitializedPlugins(t *testing.T) {
 	})
 }
 
-func TestLoader_Load_UseAPIForManifestPublicKey(t *testing.T) {
```
Comment (andresmgot) [inline pkg/plugins/manager/loader/loader_test.go:1113]: this test is no longer useful because the API download is not asynchronous now

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 7]
```diff
@@ -111,20 +152,73 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+		if !shouldKeep[key] {
+			err = pmv.kv.Del(ctx, key)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Update the last updated timestamp
+	pmv.kv.GetLastUpdated(ctx)
```
Comment (DanCech) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:187]:
```suggestion
	pmv.kv.SetLastUpdated(ctx)
```

**Review by DanCech — APPROVED** [review]
LGTM aside from that one typo

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 8]
```diff
@@ -30,17 +34,70 @@ type ManifestVerifier struct {
...
+	pmv := &ManifestVerifier{
+		cfg:  cfg,
+		mlog: mlog,
+		cli:  makeHttpClient(),
+		kv:   kv,
+	}
+	if pmv.featureEnabled() {
+		go func() {
+			pmv.updateKeysPeriodically()
+		}()
```
Comment (xnyo) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:52]: We have a pattern for background services in Grafana :smile:  . Can we use that one here instead?

```go
func (pmv *ManifestVerifier) Run(ctx context.Context) error {
	if !pmv.featureEnabled() {
		<-ctx.Done()
		return ctx.Err()
	}

	// initial run and logic to determine first tick interval, etc like in updateKeys()

	downloadKeysTicker := time.NewTicker(nextSendInterval)
	defer downloadKeysTicker.Stop()

	select {
		case <-downloadKeysTicker.C:
			// download keys
		case <-ctx.Done():
			// ctx done, perform any cleanup here
			return ctx.Err()
	}
	return ctx.Err()
}
```

Then register the background service in `pkg/server/backgroundsvcs/background_services.go` in `ProvideBackgroundServiceRegistry` and re-run wire

More info: https://github.com/grafana/grafana/blob/main/contribute/backend/services.md#background-services

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 9]
```diff
@@ -30,17 +34,70 @@ type ManifestVerifier struct {
...
+	}
+	return pmv
+}
+
+func (pmv *ManifestVerifier) featureEnabled() bool {
+	return pmv.cfg != nil && pmv.cfg.Features != nil && pmv.cfg.Features.IsEnabled("pluginsAPIManifestKey")
```
Comment (xnyo) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:58]: Nit: I would make this `"pluginsAPIManifestKey"` a const (inside `plugins` package).

There's also an instance of this literal in `loader_test.go` and `verified_test.go`

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 10]
```diff
@@ -68,40 +125,24 @@ N1c5v9v/4h6qeA==
...
 	var data struct {
 		Items []ManifestKeys
 	}
 
 	url, err := url.JoinPath(pmv.cfg.GrafanaComURL, "/api/plugins/ci/keys") // nolint:gosec URL is provided by config
 	if err != nil {
-		return "", err
+		return err
 	}
 	req, err := http.NewRequest(http.MethodGet, url, nil)
```
Comment (xnyo) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:138]: Since we now have a `context.Context` passed to this function, can we pass it to the request as well?

```go
req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
if err != nil {
	return err
}
```

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 11]
```diff
@@ -111,20 +152,73 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+		}
+	}
+
+	// Update the last updated timestamp
+	pmv.kv.GetLastUpdated(ctx)
+
+	return nil
+}
+
+// getPublicKey loads public keys from:
```
Comment (xnyo) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:192]: Nit:
```suggestion
// GetPublicKey loads public keys from:
```

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 12]
```diff
@@ -111,20 +152,73 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
+	if !pmv.featureEnabled() {
+		return publicKeyText, nil
 	}
 
-	key, exist = pmv.publicKeys[keyID]
+	pmv.lock.Lock()
+	defer pmv.lock.Unlock()
+
+	ctx := context.Background()
```
Comment (xnyo) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:203]: This now requires a `context.Context` due to the kvstore, I think it's better to now change the signature to accept a `context.Context` as first argument rather than creating a `context.Background()` here.

I have noticed the call chain for this one is a bit long until we get to a usable `context.Context`

(`GetPublicKey` -> `Signature.validateManifest` -> `Signature.readPluginManifest` -> `Signature.Calculate`)

but I would still do it, as it's the only usage apart from tests.

Alternatively, I would at least add some sort of timeout with `context.WithTimeout`, but in my opinion it's worth passing the whole context around.

File: pkg/services/pluginsintegration/keystore/keystore.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func (s *Service) Del(ctx context.Context, key string) error {
+	return s.kv.Del(ctx, prefix+key)
+}
+
+func (s *Service) GetLastUpdated(ctx context.Context) time.Time {
+	lastUpdated := time.Time{}
+	if val, ok, err := s.kv.Get(ctx, lastUpdatedKey); err != nil {
+		s.slog.Error("Failed to get last sent time", "error", err)
```
Comment (xnyo) [inline pkg/services/pluginsintegration/keystore/keystore.go:44]: Nit: I would return an error in this function rather than logging, so the caller can distinguish between an underlying failure and "job never run"

```go
lastUpdated := time.Time{}
val, ok, err := s.kv.Get(ctx, lastUpdatedKey)
if err != nil {
	return lastUpdated, fmt.Errorf("get: %w")
}
if !ok {
	// return zero-value time.Time
	return lastUpdated, nil
}
```
And I would probably treat the `time.Parse` failure as when the key does not exist (`return time.Time{}, nil`)

But interested in hearing what other folks think as well :grinning:

File: pkg/services/pluginsintegration/keystore/keystore.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+			lastUpdated = parsed
+		}
+	}
+	return lastUpdated
+}
+
+func (s *Service) SetLastUpdated(ctx context.Context) {
+	lastUpdated := time.Now()
+	if err := s.kv.Set(ctx, lastUpdatedKey, lastUpdated.Format(time.RFC3339)); err != nil {
+		s.slog.Warn("Failed to update last sent time", "error", err)
```
Comment (xnyo) [inline pkg/services/pluginsintegration/keystore/keystore.go:58]: Same here about error handling, I would personally keep the error and leave the handling to the caller

**Review by xnyo — COMMENTED** [review]
Awesome work! :rocket::rocket:
I have left some comments but it's looking great so far!

And another thing which I was not sure where to put but it's more like a question:

I am not 100% sure how Wire's magic works, but since fetching the keys now happens in the background, could we have a situation where Grafana is loading plugins during startup, but the keys haven't been fetched yet, so some plugins fail to load because the keys are not up to date?

In other words, do we have to make sure that the loader starts after the keys have been fetched?

File: pkg/plugins/manager/signature/manifestverifier/verifier.go (hunk 11 above)
Comment (wbrowne) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go:192]: It looks like this no longer needs to be exported actually 🤔

**Comment by andresmgot** [comment]
> I am not 100% sure how Wire's magic works, but since fetching the keys now happens in the background, could we have a situation where Grafana is loading plugins during startup, but the keys haven't been fetched yet, so some plugins fail to load because the keys are not up to date?

Yes, it's not guaranteed but the first load will probably use the hardcoded key (unless the load process is slow enough that by the time it `Verify`s the first plugin, the API key has been downloaded already). In the case the hardcoded key is not valid to validate the plugins, the process will fail but after a restart it will succeed. This is more like an operator pattern.

> In other words, do we have to make sure that the loader starts after the keys have been fetched?

Nope, that's the point of this PR, to not block the loader because of the the fetch.

I'm going to review the inline comments but wanted to clarify this first :) Thanks for the review!

**Review by wbrowne — COMMENTED** [review]
Just wanted to leave a quick note as I just tried it out to see what it's like to manage this from an API perspective.

To me it's a bit confusing that if you are creating the service by hand, it's not clear that kv isn't strictly necessary for the service to function as it's usage is wrapped behind a toggle. I was imagining separate implementations where we can provide one that fetches keys remotely and another that is static, so it's clearer where the key is coming from rather than mixing both.

```go
manifestverifier.New(cfg, mlog, defaultStaticKeyStore) // or manifestverifier.Default(cfg, mlog) 
```
```go
manifestverifier.New(cfg, mlog, remoteKeyStore) 
```

Maybe this is much trickier though. I'll have a think about it too

**Comment by andresmgot** [comment]
> it's not clear that kv isn't strictly necessary for the service

well, it will be once the feature flag is removed. Is it worth to optimize for the temporary code?

**Comment by wbrowne** [comment]
> > it's not clear that kv isn't strictly necessary for the service
> 
> well, it will be once the feature flag is removed. Is it worth to optimize for the temporary code?

Will there be a way to always run with just static? Else, it's fine for now and maybe we figure out how to make it easy for consumers to use similar GCOM API fetch logic

**Comment by andresmgot** [comment]
> Will there be a way to always run with just static? Else, it's fine for now and maybe we figure out how to make it easy for consumers to use similar GCOM API fetch logic

I didn't think about it. My plan was to add a way to avoid trying to download the keys, but that would still use the kvstore, which is not really necessary. I think I need to split this service (manifestverifier) in two, one for key retrieval (and this one can be completely static or dynamic) and another one for doing the verification :thinking:. I can do that as a follow-up PR, to avoid adding more changes here.

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 15]
```diff
@@ -30,18 +37,82 @@ type ManifestVerifier struct {
...
 }
 
-func New(cfg *config.Cfg, mlog log.Logger) *ManifestVerifier {
-	return &ManifestVerifier{
-		cfg:        cfg,
-		publicKeys: map[string]ManifestKeys{},
-		mlog:       mlog,
-		cli:        makeHttpClient(),
+func (pmv *ManifestVerifier) Run(ctx context.Context) error {
+	if !pmv.featureEnabled() {
```
Comment (marefr) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:56]: We should probably implement https://github.com/grafana/grafana/blob/main/contribute/backend/services.md#disabled-services. But since the registered background service is not this service, rather the `signature.Signature` maybe we should implement `service.IsDisabled ` for `signature.Signature` and remove this feature toggle check?

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 16]
```diff
@@ -68,40 +139,24 @@ N1c5v9v/4h6qeA==
...
-	}
-
-	// Retrieve the key from the API and store it in the database
+// Retrieve the key from the API and store it in the database
+func (pmv *ManifestVerifier) downloadKeys(ctx context.Context) error {
 	var data struct {
 		Items []ManifestKeys
 	}
 
 	url, err := url.JoinPath(pmv.cfg.GrafanaComURL, "/api/plugins/ci/keys") // nolint:gosec URL is provided by config
```
Comment (marefr) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:148]: Random thought. Might be time to create a proper gcom client to be used within Grafana?

File: pkg/plugins/manager/signature/manifestverifier/verifier.go [hunk 17]
```diff
@@ -111,27 +166,77 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
...
+//   - A cached value from kv storage if it has been already retrieved. This cache is populated from the grafana.com API.
+func (pmv *ManifestVerifier) getPublicKey(ctx context.Context, keyID string) (string, error) {
+	if !pmv.featureEnabled() {
+		return publicKeyText, nil
+	}
+
+	pmv.lock.Lock()
+	defer pmv.lock.Unlock()
+
+	keys, err := pmv.kv.ListKeys(ctx)
```
Comment (marefr) [inline pkg/plugins/manager/signature/manifestverifier/verifier.go:215]: Nit. Future opportunity to cache this rather than fetching the keys over and over again for each plugin being loaded.

File: pkg/plugins/ifaces.go [hunk 18]
```diff
@@ -148,3 +148,12 @@ type FeatureToggles interface {
 type SignatureCalculator interface {
 	Calculate(ctx context.Context, src PluginSource, plugin FoundPlugin) (Signature, error)
 }
+
+type KeyStore interface {
```
Comment (marefr) [inline pkg/plugins/ifaces.go:152]: Question/nit. Do we see any need of storing additional data besides the key in the store? A key being revoked was discussed in the comments, but maybe a missing key now implicitely means that? In that case or future needs (premature optimizations), might make sense to store a JSON blob instead of a string and use/return struct for updated/retrieving keys. With storing the string now would be hard to change later. Any thoughts?

**Review by marefr — APPROVED** [review]
LGTM great work. Some nits/comments/questions

File: pkg/plugins/manager/signature/manifestverifier/verifier.go (hunk 16 above)
Comment (wbrowne) [reply pkg/plugins/manager/signature/manifestverifier/verifier.go:148]: YES. I was experimenting with trying to re-use some of the existing install stuff outside of Grafana and it feels like that would be a nice improvement. An issue for another day
